libafl_frida/src/coverage_rt.rs

```diff
@@ -61,128 +142,318 @@ impl CoverageRuntime {
-            map: [0_u8; MAP_SIZE],
-            previous_pc: 0,
-            _pinned: PhantomPinned,
-        })))
+        Self {
+            inner: Rc::pin(RefCell::new(CoverageRuntimeInner {
+                map: [0_u8; MAP_SIZE],
+                previous_pc: 0,
+                _pinned: PhantomPinned,
+            })),
+            inner_bbs: Rc::pin(RefCell::new(CoverageRuntimeInner {
+                map: [0_u8; MAP_SIZE],
+                previous_pc: 0,
+                _pinned: PhantomPinned,
+            })),
+            ranges: RangeMap::new(),
+            save_dr_cov: false,
+            coverage_directory: PathBuf::from("./coverage"),
+            basic_blocks: HashMap::new(),
+            cnt: 0,
+            max_cnt: 0,
+            stored_cnt: 0,
+        }
+    }
+
+    /// Set whether to save `DrCov` files
+    #[must_use]
+    pub fn save_dr_cov(mut self, save_dr_cov: bool) -> Self {
+        self.save_dr_cov = save_dr_cov;
+        self
+    }
+
+    /// Set the coverage directory
+    #[must_use]
+    pub fn coverage_directory(mut self, coverage_directory: &str) -> Self {
+        self.coverage_directory = PathBuf::from(coverage_directory);
+        self
+    }
+
+    /// Set the maximum number of executions to accumulate before writing the coverage to disk
+    #[must_use]
+    pub fn max_cnt(mut self, max_cnt: usize) -> Self {
+        self.max_cnt = max_cnt;
+        self
+    }
+
+    /// Builder pattern
+    #[must_use]
+    pub fn build(self) -> CoverageRuntime {
+        CoverageRuntime {
+            inner: Rc::pin(RefCell::new(CoverageRuntimeInner {
+                map: [0_u8; MAP_SIZE],
+                previous_pc: 0,
+                _pinned: PhantomPinned,
+            })),
+            inner_bbs: Rc::pin(RefCell::new(CoverageRuntimeInner {
+                map: [0_u8; MAP_SIZE],
+                previous_pc: 0,
+                _pinned: PhantomPinned,
+            })),
+            ranges: self.ranges,
+            save_dr_cov: self.save_dr_cov,
+            coverage_directory: self.coverage_directory,
+            basic_blocks: self.basic_blocks,
+            cnt: self.cnt,
+            max_cnt: self.max_cnt,
+            stored_cnt: self.stored_cnt,
+        }
     }
 
     /// Retrieve the coverage map pointer
     pub fn map_mut_ptr(&mut self) -> *mut u8 {
-        self.0.borrow_mut().map.as_mut_ptr()
+        self.inner.borrow_mut().map.as_mut_ptr()
     }
 
     /// A minimal `maybe_log` implementation. We insert this into the transformed instruction stream
     /// every time we need a copy that is within a direct branch of the start of the transformed basic
     /// block.
     #[cfg(target_arch = "aarch64")]
     #[allow(clippy::cast_possible_wrap)]
-    pub fn generate_inline_code(&mut self, h64: u64) -> Box<[u8]> {
-        let mut borrow = self.0.borrow_mut();
+    pub fn generate_inline_code(&mut self, h64: u64, save_block_cov: bool) -> Box<[u8]> {
```

Comment:
this assembly code needs optimizing.

Reply:
Could you be more specific? I'm not an expert in assembly optimization